File: shell-examples/src/main/java/io/vertx/example/shell/echokeyboard/EchoKeyboardCommand.java
package io.vertx.example.shell.echokeyboard;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Launcher;
import io.vertx.core.VertxOptions;
import io.vertx.ext.dropwizard.DropwizardMetricsOptions;
import io.vertx.ext.shell.ShellService;
import io.vertx.ext.shell.ShellServiceOptions;
import io.vertx.ext.shell.command.Command;
import io.vertx.ext.shell.command.CommandBuilder;
import io.vertx.ext.shell.command.CommandRegistry;
import io.vertx.ext.shell.term.TelnetTermOptions;

/**
 * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
 */
public class EchoKeyboardCommand extends AbstractVerticle {

  public static void main(String[] args) {
    Launcher launcher = new Launcher() {
      @Override
      public void beforeStartingVertx(VertxOptions options) {
        options.setMetricsOptions(new DropwizardMetricsOptions().setEnabled(true));
      }
    };
    launcher.dispatch(new String[]{"run", EchoKeyboardCommand.class.getName()});
  }

  @Override
  public void start() throws Exception {

    Command starwars = CommandBuilder.command("echokeyboard").
        processHandler(process -> {

          // Echo
          process.stdinHandler(keys -> {
            process.write(keys.replace('\r', '\n'));
          });

          // Terminate when user hits Ctrl-C
          process.interruptHandler(v -> {
            process.end();
          });

        }).build(vertx);

    ShellService service = ShellService.create(vertx, new ShellServiceOptions().setTelnetOptions(
        new TelnetTermOptions().setHost("localhost").setPort(3000)
    ));
    CommandRegistry.getShared(vertx).registerCommand(starwars);
    service.start(ar -> {
      if (!ar.succeeded()) {
        ar.cause().printStackTrace();
      }
    });
  }
}

File: shell-examples/src/main/kotlin/io/vertx/example/shell/echokeyboard/EchoKeyboardCommand.kt
package io.vertx.example.shell.echokeyboard

import io.vertx.ext.shell.ShellService
import io.vertx.ext.shell.ShellServiceOptions
import io.vertx.ext.shell.command.CommandBuilder
import io.vertx.ext.shell.command.CommandRegistry
import io.vertx.ext.shell.term.TelnetTermOptions
import io.vertx.kotlin.ext.shell.*
import io.vertx.kotlin.ext.shell.term.*

class EchoKeyboardCommand : io.vertx.core.AbstractVerticle()  {
  override fun start() {

    var starwars = CommandBuilder.command("echokeyboard").processHandler({ process ->

      // Echo
      process.stdinHandler({ keys ->
        process.write(keys.replace('\r', '\n'))
      })

      // Terminate when user hits Ctrl-C
      process.interruptHandler({ v ->
        process.end()
      })

    }).build(vertx)

    var service = ShellService.create(vertx, ShellServiceOptions(
      telnetOptions = TelnetTermOptions(
        host = "localhost",
        port = 3000)))
    CommandRegistry.getShared(vertx).registerCommand(starwars)
    service.start({ ar ->
      if (!ar.succeeded()) {
        ar.cause().printStackTrace()
      }
    })
  }
}
